### PR TITLE
Fix: two .dist-info directories when using --build-number

### LINKED ISSUE #263: bdist_wheel --build BUILD_NUMBER creates two *.dist-info directories
**Steps to reproduce**:

From the root of the repository (e.g: this very repository) run:
```
python setup.py bdist_wheel --build-number 123
unzip -l dist/wheel-0.32.1-123-py2.py3-none-any.whl | grep dist
```

**Result**:
```
Archive:  dist/wheel-0.32.1-123-py2.py3-none-any.whl
    14481  2018-10-04 08:38   wheel/bdist_wheel.py
     1125  2018-10-04 08:42   wheel-0.32.1-123.dist-info/LICENSE.txt
     2080  2018-10-04 08:42   wheel-0.32.1-123.dist-info/METADATA
      121  2018-10-04 08:42   wheel-0.32.1-123.dist-info/WHEEL
      108  2018-10-04 08:42   wheel-0.32.1-123.dist-info/entry_points.txt
        6  2018-10-04 08:42   wheel-0.32.1-123.dist-info/top_level.txt
     1468  2018-10-04 08:42   wheel-0.32.1.dist-info/RECORD
```
Notice `RECORD` is in the wrong directory.
Trying to install this wheel:

```
$ pip install dist/*
...
Processing ./dist/wheel-0.32.1-123-py2.py3-none-any.whl
Installing collected packages: wheel
  Found existing installation: wheel 0.30.0
    Uninstalling wheel-0.30.0:
      Successfully uninstalled wheel-0.30.0
  Rolling back uninstall of wheel
Exception:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/pip/_internal/basecommand.py", line 141, in main
    status = self.run(options, args)
  File "/usr/local/lib/python2.7/dist-packages/pip/_internal/commands/install.py", line 347, in run
    use_user_site=options.use_user_site,
  File "/usr/local/lib/python2.7/dist-packages/pip/_internal/req/__init__.py", line 49, in install_given_reqs
    **kwargs
  File "/usr/local/lib/python2.7/dist-packages/pip/_internal/req/req_install.py", line 941, in install
    use_user_site=use_user_site, pycompile=pycompile,
  File "/usr/local/lib/python2.7/dist-packages/pip/_internal/req/req_install.py", line 526, in move_wheel_files
    warn_script_location=warn_script_location,
  File "/usr/local/lib/python2.7/dist-packages/pip/_internal/wheel.py", line 324, in move_wheel_files
    clobber(source, lib_dir, True)
  File "/usr/local/lib/python2.7/dist-packages/pip/_internal/wheel.py", line 273, in clobber
    ', '.join(info_dir))
AssertionError: Multiple .dist-info directories: /usr/local/lib/python2.7/dist-packages/wheel-0.32.1.dist-info, /usr/local/lib/python2.7/dist-packages/wheel-0.32.1-123.dist-info
```
**Expected result:**
```
$ unzip -l dist/wheel-0.32.1-123-py2.py3-none-any.whl | grep dist
Archive:  dist/wheel-0.32.1-123-py2.py3-none-any.whl
    14481  2018-10-04 08:38   wheel/bdist_wheel.py
     1125  2018-10-04 08:42   wheel-0.32.1-123.dist-info/LICENSE.txt
     2080  2018-10-04 08:42   wheel-0.32.1-123.dist-info/METADATA
      121  2018-10-04 08:42   wheel-0.32.1-123.dist-info/WHEEL
      108  2018-10-04 08:42   wheel-0.32.1-123.dist-info/entry_points.txt
        6  2018-10-04 08:42   wheel-0.32.1-123.dist-info/top_level.txt
     1468  2018-10-04 08:42   wheel-0.32.1-123.dist-info/RECORD
```

### PR DESCRIPTION
**Pull request by roee-allegro**
Fixes #263

### CODE REVIEW

**Comment by benoit-pierre** [comment]
How about this instead:
```diff
 wheel/wheelfile.py | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git i/wheel/wheelfile.py w/wheel/wheelfile.py
index d902d64..de54d21 100644
--- i/wheel/wheelfile.py
+++ w/wheel/wheelfile.py
@@ -41,7 +41,9 @@ def __init__(self, file, mode='r'):
 
         super(WheelFile, self).__init__(file, mode, compression=ZIP_DEFLATED, allowZip64=True)
 
-        self.dist_info_path = '{}.dist-info'.format(self.parsed_filename.group('namever'))
+        self.dist_info_path = '{}.dist-info'.format(
+            '-'.join(filter(bool, self.parsed_filename.group('namever', 'build')))
+        )
         self.record_path = self.dist_info_path + '/RECORD'
         self._file_hashes = OrderedDict()
         self._file_sizes = {}
```
Adding a test would be nice.

File: wheel/wheelfile.py [hunk 1]
```diff
@@ -41,7 +41,9 @@ def __init__(self, file, mode='r'):
 
         super(WheelFile, self).__init__(file, mode, compression=ZIP_DEFLATED, allowZip64=True)
 
-        self.dist_info_path = '{}.dist-info'.format(self.parsed_filename.group('namever'))
+        self.dist_info_path = '{}.dist-info'.format(
+            '-'.join(filter(bool, self.parsed_filename.group('namever', 'build')))
```
Comment (benoit-pierre) [inline wheel/wheelfile.py]: `filter(None, ...)` like you used in the test is better.

**Comment by agronholm** [comment]
Seems to me that there was a deeper design problem if the name of the dist-info directory was computed twice during the building of the wheel. I'll have to investigate that first.

**Comment by agronholm** [comment]
I'm going to wait until I get an answer for https://github.com/pypa/wheel/issues/263#issuecomment-427124643 before committing any fix.

**Comment by agronholm** [comment]
Apparently the build number was never supposed to appear in the `.dist-info` directory name, so I'm going to have to reject this and fix the `bdist_wheel` command instead. Thanks for the effort nonetheless.

**Pull request closed without merging**